File: IrrgartenRuby/irrgarten/game.rb
# encoding: utf-8

require_relative 'dice'
require_relative 'game_character'
require_relative 'game_state'
require_relative 'monster'
require_relative 'player'
require_relative 'labyrinth'
require_relative 'fuzzy_player'

module Irrgarten
       class Game
	  @@MAX_ROUNDS = 10 #Es de clase pero también constante

	  def initialize(n_players)
	      @current_player_index=Dice.who_starts(n_players)
	      @log=""
	      @players=Array.new
	      @monsters=Array.new
	      n_players.times do |i|
			@players << Player.new(i, Dice.random_intelligence, Dice.random_strength)
		  end

		  n_rows = 10
          n_cols = 10
          exit_row = 4
          exit_col = 3
	      @labyrinth=Labyrinth.new(n_rows, n_cols, exit_row, exit_col)
	      @current_player=@players[@current_player_index]
	      configure_labyrinth
	  end

	  def finished()
	    @labyrinth.have_a_winner
	  end

	  def next_step(preferred_direction)
	      @log=""
	      dead=@current_player.dead()
	      if !dead
			direction=actual_direction(preferred_direction)

			if (direction != preferred_direction)
			log_player_no_orders()
			end
			monster=@labyrinth.put_player(direction,@current_player)

			if monster == nil
			log_no_monster()
			else
			winner=combat(monster)
			manage_reward(winner)
			end
	      else
			manage_resurrection()
	      end
	      endgame=finished()
	      if !endgame
			next_player()
	      end
	      endgame
	  end

	  def game_state()
		player=""
	    monster=""

	    (0..@players.length).each do |i|
			player += @players[i].to_s
	    end

	    (0..@monsters.length).each do |j|
			monster += @monsters[j].to_s
	    end

	    GameState.new(@labyrinth.to_s,player,monster,@current_player_index.to_s,
	      @labyrinth.have_a_winner,@log)
	  end

	  private

	  def configure_labyrinth()
		@labyrinth.add_block(Orientation::HORIZONTAL, 0, 0, 9);
		@labyrinth.add_block(Orientation::HORIZONTAL, 8, 1, 7);
		@labyrinth.add_block(Orientation::HORIZONTAL, 5, 1, 2);
		@labyrinth.add_block(Orientation::HORIZONTAL, 6, 2, 3);
		@labyrinth.add_block(Orientation::HORIZONTAL, 2, 4, 2);
		@labyrinth.add_block(Orientation::HORIZONTAL, 4, 4, 2);
		@labyrinth.add_block(Orientation::VERTICAL, 1, 0, 5);
		@labyrinth.add_block(Orientation::VERTICAL, 2, 2, 3);
		@labyrinth.add_block(Orientation::VERTICAL, 1, 8, 6);
		@labyrinth.add_block(Orientation::VERTICAL, 2, 6, 7);
	   
		monstruo1=Monster.new("MONSTRUO1",100,100)
	    monstruo2=Monster.new("MONSTRUO2",100,100)
	    monstruo3=Monster.new("MONSTRUO3",100,100)

		@labyrinth.add_monster(6,1,monstruo1)
	    @labyrinth.add_monster(2,3,monstruo2)
		@labyrinth.add_monster(7,7,monstruo3)
	    @monsters << monstruo1
	    @monsters << monstruo2
		@monsters << monstruo3

		@labyrinth.spread_players(@players)
	  end

	  def next_player()
	    if @current_player_index == @players.size-1
		@current_player_index=0
	      else
		@current_player_index+=1
	      end
	      @current_player=@players[@current_player_index]
	  end

	  def actual_direction(preferred_direction)
	      current_row=@current_player.row
	      current_col=@current_player.col
	      valid_moves=@labyrinth.valid_moves(current_row,current_col)
	      output=@current_player.move(preferred_direction,valid_moves)
	      output
	  end

	  def combat(monster)
	      rounds=0
	      winner=GameCharacter::PLAYER
	      player_attack=@current_player.attack
	      lose=monster.defend(player_attack)
	      
	      while (!lose && rounds<@@MAX_ROUNDS)
			winner=GameCharacter::MONSTER
			rounds+=1
			monsterattack=monster.attack
			lose=@current_player.defend(monsterattack)

			if !lose
			playerattack=@current_player.attack
			winner=GameCharacter::PLAYER
			lose=monster.defend(playerattack)
			end
	      end

	      log_rounds(rounds,@@MAX_ROUNDS)
	      winner
	  end

	  def manage_reward(winner)
	    if winner==GameCharacter::PLAYER
			@current_player.receive_reward()
			log_player_won()
	    else
			log_monster_won()
	    end
	  end

	  def manage_resurrection()
	      resurrect = Dice.resurrect_player()
	      if resurrect
			@current_player.resurrect()
			fuzzy_player = FuzzyPlayer.new(@current_player)
			@players[@current_player_index] = fuzzy_player
			@labyrinth.set_fuzzy_player(@current_player.row, @current_player.col, fuzzy_player)
			@current_player = fuzzy_player
			log_resurrected()
	      else
			log_player_skip_turn()
	      end
	  end

	  def log_player_won()
	      @log=@log.to_s+" El jugador ha ganado el combate\n"
	  end

	  def log_monster_won()
	    @log=@log.to_s+" El monstruo ha ganado el combate\n"
	  end

	  def log_resurrected()
	    @log=@log.to_s+" El jugador ha resucitado\n"
	  end

	  def log_player_skip_turn()
	    @log=@log.to_s+" El jugador ha perdido el turno\n"
	  end

	  def log_player_no_orders()
	    @log=@log.to_s+" No se han seguido las instrucciones\n"
	  end

	  def log_no_monster()
	    @log=@log.to_s+ " No ha sido posible moverse o se ha movido a una celda vacía\n"
	  end

	  def log_rounds(rounds, max)
	    @log=@log.to_s+ " Se han producido un total de "+rounds.to_s+" rondas de "+max.to_s+" maximas rondas de combate\n"
	  end
	end


end

File: IrrgartenJava/IrrgartenJava/src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;

/**
 * La clase Game representa un juego en un laberinto con jugadores y monstruos.
 * Proporciona métodos para gestionar el progreso del juego, como avanzar a la siguiente etapa,
 * verificar si el juego ha terminado y obtener el estado actual del juego.
 * Además, incluye métodos internos para configurar el laberinto, manejar turnos de jugadores, 
 * gestionar combates y recompensas, y llevar un registro de eventos relevantes del juego.
 * 
 * @author Juan Carlos Vílchez Sánchez
 */
public class Game {

    // Constante de la clase
    private static final int MAX_ROUNDS = 10;

    // Variables de instancia
    private int currentPlayerIndex;
    private String log;
    private Player currentPlayer;
    private final Labyrinth labyrinth;
    private final ArrayList<Monster> monsters;
    private final ArrayList<Player> players;

    /**
     * Construye un nuevo juego con el número especificado de jugadores.
     *
     * @param nPlayers El número de jugadores en el juego.
     */
    public Game(int nPlayers) {
        // Inicialización de variables
        log = "";
        currentPlayerIndex = Dice.whoStarts(nPlayers);
        players = new ArrayList<>();
        monsters = new ArrayList<>();

        // Creación de jugadores
        for (int i = 0; i < nPlayers; i++) {
            Player jugador = new Player((char) ('0' + i), Dice.randomIntelligence(), 
                             Dice.randomStrength());
            players.add(jugador);
        }
        currentPlayer = players.get(currentPlayerIndex);

        // Creación y configuración del laberinto
        int nRows = 10;
        int nCols = 10;
        int exitRow = 4;
        int exitCol = 3;
        labyrinth = new Labyrinth(nRows, nCols, exitRow, exitCol); 
        configureLabyrinth();
    }

    /**
     * Verifica si el juego ha terminado.
     *
     * @return true si el juego ha terminado, false en caso contrario.
     */
    public boolean finished() {
        return labyrinth.haveAWinner();
    }

    /**
     * Realiza la siguiente acción en el juego, basada en la dirección preferida por el jugador actual.
     *
     * @param preferredDirection La dirección preferida por el jugador actual.
     * @return true si la acción se realizó con éxito, false en caso contrario.
     */
    public boolean nextStep(Directions preferredDirection) {
        log = "";
        if(!currentPlayer.dead()){
            Directions direction = actualDirection(preferredDirection);
            
            if(direction != preferredDirection){
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if(monster == null){
                logNoMonster();
            }else{
                GameCharacter winner = combat(monster);
                manageReward(winner);
            } 
        }else{
            manageResurrection();
        }
        boolean endGame = finished();
        
        if(!endGame){
            nextPlayer();
        }
        return endGame;
    }

    /**
     * Obtiene el estado actual del juego.
     *
     * @return El estado actual del juego encapsulado en un objeto GameState.
     */
    public GameState getGameState() {       
        String namesplayers = "";
        String namesmonsters = "";
        for(Player player : players){
            namesplayers += player.toString();
        }
        for(Monster monster : monsters){
            namesmonsters += monster.toString();
        }
        return new GameState(labyrinth.toString(), namesplayers, namesmonsters,
                currentPlayerIndex, this.finished(),log);
    }

    // Métodos privados

    /**
     * Configura el laberinto, añadiendo bloques y colocando monstruos en posiciones específicas.
     */
    private void configureLabyrinth() {
        labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, 9);
        labyrinth.addBlock(Orientation.HORIZONTAL, 8, 1, 7);
        labyrinth.addBlock(Orientation.HORIZONTAL, 5, 1, 2);
        labyrinth.addBlock(Orientation.HORIZONTAL, 6, 2, 3);
        labyrinth.addBlock(Orientation.HORIZONTAL, 2, 4, 2);
        labyrinth.addBlock(Orientation.HORIZONTAL, 4, 4, 2);
        labyrinth.addBlock(Orientation.VERTICAL, 1, 0, 5);
        labyrinth.addBlock(Orientation.VERTICAL, 2, 2, 3);
        labyrinth.addBlock(Orientation.VERTICAL, 1, 8, 6);
        labyrinth.addBlock(Orientation.VERTICAL, 2, 6, 7);

        // Colocación de monstruos en posiciones específicas
        
        
        int newpos1[] = {6,1};
        Monster monstruo_1 = new Monster ("monstruo_1", 
                Dice.randomIntelligence(), Dice.randomStrength());
        monsters.add(monstruo_1);
        labyrinth.addMonster(newpos1[0], newpos1[1], monsters.get(0));

        int newpos2[] = {2,3};
        Monster monstruo_2 = new Monster ("monstruo_2", 
                100, 100);
        monsters.add(monstruo_2);
        labyrinth.addMonster(newpos2[0], newpos2[1], monsters.get(1));
        
        int newpos3[] = {7,7};
        Monster monstruo_3 = new Monster ("monstruo_3", 
                Dice.randomIntelligence(), Dice.randomStrength());
        monsters.add(monstruo_3);
        labyrinth.addMonster(newpos3[0], newpos3[1], monsters.get(2));
        
        labyrinth.spreadPlayers(players);
    }

    /**
     * Cambia al siguiente jugador en la secuencia de turnos.
     */
    private void nextPlayer() {
        if (currentPlayerIndex == players.size()-1){
            currentPlayerIndex = 0;
        } else {
            currentPlayerIndex += 1;
        }
        currentPlayer = players.get(currentPlayerIndex);
    }

    /**
     * Determina la dirección actual basándose en la preferencia del jugador.
     *
     * @param preferredDirection La dirección preferida por el jugador.
     * @return La dirección actualizada basada en la preferencia del jugador.
     */
    private Directions actualDirection(Directions preferredDirection) {
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);
        Directions output = currentPlayer.move(preferredDirection, validMoves);
        
        return output;
    }

    /**
     * Simula un combate entre el jugador actual y un monstruo.
     *
     * @param monster El monstruo con el que se realiza el combate.
     * @return El personaje ganador del combate.
     */
    private GameCharacter combat(Monster monster) {
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        
        
        while((!lose) && (rounds < MAX_ROUNDS)){
            rounds += 1;
            winner = GameCharacter.MONSTER;
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            
            if(!lose){
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        logRounds(rounds, MAX_ROUNDS);
        
        return winner;  
    }

    /**
     * Gestiona las recompensas después de un combate.
     *
     * @param winner El personaje que ganó el combate.
     */
    private void manageReward(GameCharacter winner) {
        if(winner == GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        }else{
            logMonsterWon();
        }
    }

    /**
     * Gestiona la resurrección de jugadores después de ciertas condiciones.
     */
    private void manageResurrection() {
        boolean resurrect = Dice.resurrectPlayer();
        
        if(resurrect){
            currentPlayer.resurrect();
            FuzzyPlayer fuzzyPlayer = new FuzzyPlayer(currentPlayer);
            players.set(currentPlayerIndex, fuzzyPlayer);
            labyrinth.setFuzzyPlayer(currentPlayer.getRow(), 
                    currentPlayer.getCol(), fuzzyPlayer);
            currentPlayer = fuzzyPlayer;
            logResurrected();
        }else{
            logPlayerSkipTurn();
        }
    }

    // Métodos de registro

    private void logPlayerWon() {
        log = log + "El jugador ha ganado el combate.\n";
    }

    private void logMonsterWon() {
        log = log + "El monstruo ha ganado el combate.\n";
    }

    private void logResurrected() {
        log = log + "El jugador ha resucitado.\n";
    }

    private void logPlayerSkipTurn() {
        log = log + "El jugador ha perdido el turno por estar muerto.\n";
    }

    private void logPlayerNoOrders() {
        log = log + "El jugador no ha seguido las órdenes del jugador humano "
                  + "(no fue posible).\n";
    }

    private void logNoMonster() {
        log = log + "El jugador se ha movido a una celda vacia o no se ha"
                  + " podido mover.\n";
    }

    private void logRounds(int rounds, int max) {
        log = log + "Se han producido un total de " + rounds 
                  + " de un máximo de " + max + "ronda/s.\n";
    }
}

